NMI registration API and using special stack

finikorg opened the pull request:
NMI registration API and using special stack

finikorg committed:
arch/x86: Add NMI registration API

Add simple NMI registration API.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
arch/x86: Use NMI stack for NMIs

NMI can be triggered at any time, even when in the process of
switching stacks. Use special stack for it.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

andrewboie commented on arch/x86/core/fatal.c:
```diff
@@ -432,3 +432,18 @@ void z_x86_do_kernel_oops(const z_arch_esf_t *esf)
+
+void z_x86_do_kernel_nmi(const z_arch_esf_t *esf)
+{
+	Z_STRUCT_SECTION_FOREACH(nmi_handler, handler) {
```
Without more information on why we need a special section with chained NMI handlers, I'd rather this was simpler, just a weak function which by default calls z_x86_fatal_error(), can be overridden at the application level.

In particular the need for multiple NMI handlers seems really unlikely to me, but if you have supporting use-cases do share. Keep in mind that these functions basically have to be their own standalone environment with no kernel services available. That includes driver APIs which use locks for synchronization. It looks suspicious to me that these handlers take a struct device as an argument...you have no idea what state the device is in

finikorg replied:
@andrewboie , I have done as you requested, to make "iterable section of callback pointers at build time", is this correct?

andrewboie replied:
this doesn't answer my question: what are the use-cases for having multiple callbacks? what are you planning to do inside these callbacks? why do we need multiple callbacks?

also in this current design, if no NMI handlers are installed the exception isn't treated as a fatal error

finikorg replied:
I need only one callback, I can make it working only for me with weak function, if somebody need another callback he would reimplement this. I will fix no callback case.

finikorg replied:
@andrewboie,  Re-implemented with weak function.

andrewboie commented on arch/x86/core/intel64/cpu.c:
```diff
@@ -28,6 +28,11 @@ extern uint8_t _exception_stack1[];
 extern uint8_t _exception_stack2[];
 extern uint8_t _exception_stack3[];
 
+extern uint8_t _exception_stack4[];
```
please don't call these 4,5,6,7 this is confusing, call them z_x86_nmi_stack or something like that

finikorg replied:
renamed

andrewboie commented on arch/x86/core/intel64/locore.S:
```diff
@@ -820,16 +820,18 @@ IRQ(248); IRQ(249); IRQ(250); IRQ(251); IRQ(252); IRQ(253); IRQ(254); IRQ(255)
+#define NMI_STACK	2
 #else
 #define	IRQ_STACK	1
+#define NMI_STACK	6 /* NMI stack */
```
this doesn't work if CONFIG_X86_KPTI is enabled. You will need logic in `except` to switch to the per-cpu NMI stack manually. One of the reasons why I think this needs tests, it would be very good to somehow trigger an NMI on a CPU that is running a user thread (perhaps set a DPL of 3 on the NMI exception if CONFIG_TEST is turned on, and trigger from user mode with an `int` instruction, that might work although I am not 100% sure)

finikorg replied:
@andrewboie, do you mean `except` in `arch/x86/core/intel64/locore.S`? I have not found how other stacks are handled there.

andrewboie replied:
yes. that code contains a manual stack switch if KPTI is turned on.

finikorg replied:
Fixed for KPTI case also in the commit 43b0ae4d1bea9626c70a6e4218f6aaf1175fdb3d

andrewboie commented on arch/x86/core/intel64/locore.S:
```diff
@@ -820,16 +820,18 @@ IRQ(248); IRQ(249); IRQ(250); IRQ(251); IRQ(252); IRQ(253); IRQ(254); IRQ(255)
+#define NMI_STACK	6 /* NMI stack */
 #define EXC_STACK	7
-#define BAD_STACK	7 /* Horrible things: NMIs, double faults, MCEs */
+#define BAD_STACK	7 /* Horrible things: double faults, MCEs */
```
is it possible to use the NMI stack for these as well? can an MCE fire while an NMI is being handled? a double fault is always a pathological error.

finikorg replied:
Do you mean make BAD_STACK = NMI_STACK = 6 ? I can change to this.
Then you asked to name arrays as `z_x86_nmi_stackN`, probably I need better name.

tyrantlizard77 reviewed (commented):
Comments inline. I think this also perhaps needs a test case under tests/arch/x86 to get some coverage of these code paths.

andrewboie commented on arch/x86/core/fatal.c:
```diff
@@ -432,3 +432,18 @@ void z_x86_do_kernel_oops(const z_arch_esf_t *esf)
 }
+
+/**
+ * NMI handling
```
unless you plan on implementing this for 32-bit as well I think we should move this to x86_64 only-code

finikorg replied:
moved to x86_64

finikorg committed:
tests: nmi: Add NMI registration API test

Test that NMI registration works as expected.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
arch: x86_64: Using right exception stack with KPTI

With kernel page table isolation (KPTI) we cannot use right exception
stack since after using trampoline stack there was always switch to
7th IST stack (__x86_tss64_t_ist7_OFFSET). Make this configurable as a
parameter in EXCEPT(nr, ist) and EXCEPT_CODE(nr, ist). For the NMI we
would use ist6 (_nmi_stack).

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
arch: x86_64: Trivial correction

Correct register name in comment.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
tests: nmi: Add test verifying right stack switched to

Verify that we use right nmi_stack[] in exception.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg commented:
recheck

andrewboie commented on arch/x86/core/intel64/locore.S:
```diff
@@ -476,17 +487,17 @@ except: /*
 	movq %rsp, %r11
 
 	/* Switch to the exception stack */
```
please update this comment

finikorg replied:
done

andrewboie commented on arch/x86/core/intel64/locore.S:
```diff
@@ -562,6 +573,21 @@ except: /*
 	iretq
 
+#ifdef CONFIG_X86_KPTI
+EXCEPT      ( 0, 7); EXCEPT      ( 1, 7); EXCEPT       (2, 6); EXCEPT      ( 3, 7)
```
can we macro the second argument for clarity, EXC_STACK, NMI_STACK, etc

finikorg replied:
@andrewboie if I use macro names like EXC_STACK, NMI_STACK they are not expanded, this code (https://github.com/finikorg/zephyr/blob/up-clean/arch/x86/core/intel64/locore.S#L434) produces
`__x86_tss64_t_istEXC_STACK_OFFSET` instead of `__x86_tss64_t_ist7_OFFSET`

andrewboie commented on arch/x86/core/intel64/cpu.c:
```diff
@@ -28,6 +28,11 @@ extern uint8_t _exception_stack1[];
 extern uint8_t _exception_stack2[];
 extern uint8_t _exception_stack3[];
 
+extern uint8_t _nmi_stack[];
```
please prefix with `z_x86`, `_` prefix is deprecated (MISRA-C violation)

finikorg replied:
Fixed for `_nmi_stack` as well for `_exception_stack`

tyrantlizard77 reviewed (commented):
minor stuff

finikorg committed:
arch: x86_64: Rename _nmi_stack to z_x86_nmi_stack

Rename stack name according to MISRA-C standard.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
arch: x86_64: Rename _exception_stack to z_x86_exception_stack

Rename stack name according to MISRA-C standard.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>
